Userland/Libraries/LibAccelGfx/Painter.cpp

```diff
@@ -208,6 +235,139 @@ void Painter::draw_scaled_bitmap(Gfx::FloatRect const& dst_rect, Gfx::Bitmap con
+    auto glyphs_texure_painter = Gfx::Painter(*glyphs_texture_bitmap);
+    for (auto const& [glyphs_texture_key, glyph_bitmap] : glyph_bitmaps) {
+        auto rect = m_glyphs_texture_map.get(glyphs_texture_key).value();
+        glyphs_texure_painter.blit({ rect.x(), rect.y() }, glyph_bitmap, glyph_bitmap->rect());
+    }
+
+    m_glyphs_texture_size = glyphs_texture_bitmap->size();
+
+    glBindTexture(GL_TEXTURE_2D, m_glyphs_texture);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, glyphs_texture_bitmap->width(), glyphs_texture_bitmap->height(), 0, GL_BGRA, GL_UNSIGNED_BYTE, glyphs_texture_bitmap->scanline(0));
+}
+
+void Painter::draw_glyph_run(Vector<Gfx::DrawGlyphOrEmoji> const& glyph_run, Color const& color)
+{
+    Vector<GLfloat> vertices;
+
+    for (auto& glyph_or_emoji : glyph_run) {
+        if (glyph_or_emoji.has<Gfx::DrawGlyph>()) {
+            auto& glyph = glyph_or_emoji.get<Gfx::DrawGlyph>();
+
+            auto const* font = glyph.font;
+            auto code_point = glyph.code_point;
+            auto point = glyph.point;
+
+            auto maybe_texture_rect = m_glyphs_texture_map.get(GlyphsTextureKey { font, code_point });
+            VERIFY(maybe_texture_rect.has_value());
+
+            auto texture_rect = to_texture_space(maybe_texture_rect.value().to_type<float>(), m_glyphs_texture_size);
+
+            auto rect = Gfx::FloatRect { point + Gfx::FloatPoint(font->glyph_left_bearing(code_point), 0), maybe_texture_rect->size().to_type<float>() };
+            auto rect_in_clip_space = to_clip_space(rect);
+
+            auto p0 = rect_in_clip_space.top_left();
+            auto p1 = rect_in_clip_space.top_right();
+            auto p2 = rect_in_clip_space.bottom_left();
+            auto p3 = rect_in_clip_space.bottom_right();
+
+            auto s0 = texture_rect.top_left();
+            auto s1 = texture_rect.top_right();
+            auto s2 = texture_rect.bottom_left();
+            auto s3 = texture_rect.bottom_right();
+
+            vertices.append(p0.x());
+            vertices.append(p0.y());
+            vertices.append(s0.x());
+            vertices.append(s0.y());
+            vertices.append(p1.x());
+            vertices.append(p1.y());
+            vertices.append(s1.x());
+            vertices.append(s1.y());
+            vertices.append(p3.x());
+            vertices.append(p3.y());
+            vertices.append(s3.x());
+            vertices.append(s3.y());
+
+            vertices.append(p0.x());
+            vertices.append(p0.y());
+            vertices.append(s0.x());
+            vertices.append(s0.y());
+            vertices.append(p3.x());
+            vertices.append(p3.y());
+            vertices.append(s3.x());
+            vertices.append(s3.y());
+            vertices.append(p2.x());
+            vertices.append(p2.y());
+            vertices.append(s2.x());
+            vertices.append(s2.y());
```

Comment:
Fly-by comment, but this seems very fragile and kinda hard to stare at. Can we do this more...programmatically? Helper lambdas or anything? The changes to`draw_scaled_bitmap` give similar vibes

Reply:
sure, I added helpers and now it looks a bit less verbose.